featuretests/application_config_test.go

```diff
@@ -0,0 +1,223 @@
+func (s *ApplicationConfigSuite) TestConfigAndConfigGetReturnAllCharmSettings(c *gc.C) {
+	// initial deploy with custom settings
+	s.assertSameConfigOutput(c, initialConfig)
+
+	// use 'juju config foo=' to change values
+	s.configCommandOutput(c, s.appName,
+		"booleandefault=false",
+		"booleannodefault=true",
+		"booleanoverwrite=true", //charm default
+		"floatdefault=7.2",
+		"floatnodefault=10.2",
+		"floatoverwrite=11.1", //charm default
+		"intdefault=22",
+		"intnodefault=11",
+		"intoverwrite=111", //charm default
+		"strdefault=not",
+		"strnodefault=maybe",
+		"stroverwrite=me",
+	)
+	s.assertSameConfigOutput(c, updatedConfig)
+
+	// 'juju config --reset' to reset settings to charm default
+	s.configCommandOutput(c, s.appName, "--reset",
+		"booleandefault,booleannodefault,booleanoverwrite,floatdefault,"+
+			"floatnodefault,floatoverwrite,intdefault,intnodefault,intoverwrite,"+
+			"strdefault,strnodefault,stroverwrite")
+	s.assertSameConfigOutput(c, resetConfig)
+}
+
+func (s *ApplicationConfigSuite) assertSameConfigOutput(c *gc.C, expectedValues settingsMap) {
+	s.assertJujuConfigOutput(c, s.configCommandOutput(c, s.appName), expectedValues)
+	s.assertHookOutput(c, s.getHookOutput(c), expectedValues)
+}
+
+func (s *ApplicationConfigSuite) assertHookOutput(c *gc.C, obtained charm.Settings, expected settingsMap) {
+	c.Assert(len(obtained), gc.Equals, len(expected))
+	c.Assert(len(obtained), gc.Equals, len(s.settingKeys))
+	for name, aSetting := range expected {
+		c.Assert(s.settingKeys.Contains(name), jc.IsTrue)
+		// due to awesome float64/int parsing confusion, it's actually safer to ensure that
+		// values' string representations match
+		c.Assert(fmt.Sprintf("%v", aSetting.Value), gc.DeepEquals, fmt.Sprintf("%v", obtained[name]))
+	}
+}
+
+func (s *ApplicationConfigSuite) assertJujuConfigOutput(c *gc.C, jujuConfigOutput string, expected settingsMap) {
+	var appSettings ApplicationSetting
+	err := yaml.Unmarshal([]byte(jujuConfigOutput), &appSettings)
+	c.Assert(err, jc.ErrorIsNil)
+	obtained := appSettings.Settings
+
+	c.Assert(len(obtained), gc.Equals, len(expected))
+	c.Assert(len(obtained), gc.Equals, len(s.settingKeys))
+	for name, aSetting := range expected {
+		c.Assert(s.settingKeys.Contains(name), jc.IsTrue)
+		c.Assert(aSetting.Default, gc.Equals, obtained[name].Default)
+		c.Assert(aSetting.Value, gc.Equals, obtained[name].Value)
+	}
+}
+
+type configSetting struct {
+	Value   interface{}
+	Default bool
+}
+
+type settingsMap map[string]configSetting
+
+var (
+	initialConfig = settingsMap{
+		"booleandefault":   {true, true},
+		"booleannodefault": {Default: true},
+		"booleanoverwrite": {false, false},
+		"floatdefault":     {4.2, true},
+		"floatnodefault":   {Default: true},
+		"floatoverwrite":   {2.1, false},
+		"intdefault":       {42, true},
+		"intnodefault":     {Default: true},
+		"intoverwrite":     {1620, false},
+		"strdefault":       {"charm default", true},
+		"strnodefault":     {Default: true},
```

Comment:
So I see that you have asserts like:
c.Assert(aSetting.Value, gc.Equals, obtained[name].Value)
But here you aren't defining a Value. What do we actually get out of this, then? I suppose with nothing set then "Value" is just 'nil'. Would it be more obvious to just set it directly to:
{nil, true}
instead of using the Zero value to do that? I feel like it would make the test setup a bit easier to parse, since it would be consistent.